### PR TITLE
consistently handle StringTypeOf overloads

### PR DESCRIPTION
**Pull request by MartinNowak**
- use auto ref to avoid expensive copies
- leave the overloads undocumented
- move them below the real implementations
- use explict template instantiation to avoid casts
- had to factor out LineSplitter b/c of fwd
  reference complaints by the compiler

### CODE REVIEW

**Comment by MartinNowak** [comment]
Diff w/o whitespace changes.
https://github.com/D-Programming-Language/phobos/pull/3758/files?w=1

File: std/string.d [hunk 1]
```diff
@@ -482,12 +475,12 @@ ptrdiff_t indexOf(Range)(Range sIn, in dchar c,
     return -1;
 }
 
-ptrdiff_t indexOf(T, size_t n)(ref T[n] s, in dchar c,
-        in CaseSensitive cs = CaseSensitive.yes) @safe pure
-    if (isSomeChar!T)
+ptrdiff_t indexOf(Range)(auto ref Range s, in dchar c,
```
Comment (MartinNowak) [inline std/string.d]: The static array case is now handled by StringTypeOf as well. I'm puzzled how someone had seriously fixed on that particular overload.

**Comment by burner** [comment]
I don't like that the overloads are undocumented. How is the user supposed to know that he can call the function. IMO if it undocumented we could remove it.

File: std/string.d (hunk 1 above)
Comment (burner) [inline std/string.d]: ´´´d
char[10] superSmallBuffer;
/// load stuff in buf
auto i =indexOf(superSmallBuffer, '\n');
```

Comment (burner) [inline std/string.d]: feels like this was used in some compiler somewhere

**Comment by DmitryOlshansky** [comment]
> I don't like that the overloads are undocumented.

On the contrary I like because they are just hacks. Everybody would think that alias this to string is as good as string.

LGTM will merge unless other objections come forward.

**Comment by burner** [comment]
I disagree, I know at least one user how is using these functions and would be surprised to see that he is successfully using a function that is actually not there (not documented).

**Comment by MartinNowak** [comment]
I don't think anybody reads those [template constraints](http://dtest.thecybershadow.net/artifact/website-00f58ff908fe63c298ce44f289ca5bc3856bbe82-af184dbcb5c94f56af79ef4eccff2a34/web/phobos-prerelease/std_string.html#.stripRight) and adding a second pair doesn't make it better IMO.
```
auto stripRight(Range)(Range str) if (isSomeString!Range || isRandomAccessRange!Range && hasLength!Range && hasSlicing!Range && isSomeChar!(ElementEncodingType!Range));
```
Let's see how it looks like.

**Comment by burner** [comment]
IMO, it is actually ok looking.

for what it is worth I think currently we have the choice between two sub-optimal options:
1. Two longish template constraints in the documentation
2. Not all public function been documented.

 IMO option 1. is the better choice

**Comment by MartinNowak** [comment]
Wasn't the consensus last time to slowly deprecate those b/c we don't want almost noop overloads.
Then not documenting them would be the correct first step.

**Comment by burner** [comment]
not that I recall. I the end these PRs are only restoring previously good state. I'm not saying that any new function have these two overloads.

**Comment by MartinNowak** [comment]
>  Or, (2) can be accomplished by overloading isDir() to accept string arguments. But this implies adding an overload to every function that accepts an InputRange. This is out of the question.

https://issues.dlang.org/show_bug.cgi?id=15027#c5
https://issues.dlang.org/show_bug.cgi?id=15027#c13

We don't want to advertise this sort of overloading, but lack a better solution for now.
Also I consider an overload overload with an almost redundant constraint in our docs severe noise (someone learning D must think we're doing sth. wrong). It also doesn't convey any semantic information, b/c alias this is like deriving, the overload is only a crutch b/c we can't yet put constraints into the type deduction.

**Comment by burner** [comment]
OK.

I'll fix the other functions similar to fix PR.

**Comment by MartinNowak** [comment]
> I'll fix the other functions similar to fix PR.

Thanks a lot for helping out with this.

> OK.

I'm a bit undecided either but lean towards cleaner documentation.

**Comment by DmitryOlshansky** [comment]
OK let's roll while we have a consensus here. Yeah not ideal but let move forward.

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled off

**Comment by DmitryOlshansky** [comment]
Now re-reading it - I'm not so certain. I'll leave it to @burner

**Comment by burner** [comment]
as the discussion concluded these overloads are there to recreate old behavior and is not to be considered idomatic D. Therefore, the overloads should not appear in the docs.

**Comment by MartinNowak** [comment]
I dropped the commit.

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by MartinNowak** [comment]
While working on std.file I found that the alias this problem occurs only with `isForwardRange` but not with `isInputRange`. So our fix works (in particular it fixes static arrays), but sometimes incurrs a slightly more expensive copy. We could use auto ref on the original function to workaround this. Other ideas?

Here is an example to illustrate the problem.
```d
import std.range, std.traits;

@property bool isDir(R)(R name)
    if (isInputRange!R && isSomeChar!(ElementEncodingType!R))
{
    pragma(msg, "isDir1 ", R);
    return false;
}

@property bool isDir(R)(auto ref R name)
    if (!(isInputRange!R && isSomeChar!(ElementEncodingType!R))
        && is(StringTypeOf!R))
{
    pragma(msg, "isDir2 ", R, __traits(isRef, name));
    return isDir!(StringTypeOf!R)(name);
}

unittest
{
    static struct DontCopyMe
    {
        string s; alias s this;
        this(this) { assert(0); }
    }
    auto s = DontCopyMe("foo");
    isDir(s);
}
```

**Comment by rainers** [comment]
Nasty.

>```
> this(this) { assert(0); }
>```

you can also use `@disable this(this);` to get a compilation error.

**Comment by rainers** [comment]
This seem to work:
```
@property bool isDir(R)(R name)
if (isInputRange!R && isSomeChar!(ElementEncodingType!R) &&
	(!is(StringTypeOf!R) || is(R == StringTypeOf!R)))
{
    pragma(msg, "isDir1 ", R);
    return false;
}

@property bool isDir(R)(auto ref R name)
if (is(StringTypeOf!R) && !is(R == StringTypeOf!R))
{
    pragma(msg, "isDir2 ", R, __traits(isRef, name));
    return isDir!(StringTypeOf!R)(name);
}
```

**Comment by rainers** [comment]
Tried this in https://github.com/D-Programming-Language/phobos/pull/3764

**Pull request merged**